### PR TITLE
[WIP] Transaction statement

### PR DESCRIPTION
**Pull request by andyfengHKU**
This PR deprecate transaction API `connection.commit()` by transaction statement `connection.query("COMMIT")`

### CODE REVIEW

File: src/include/processor/operator/transaction.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include "processor/operator/physical_operator.h"
+#include "transaction/transaction_action.h"
+
+namespace kuzu {
+namespace processor {
+
+class Transaction : public PhysicalOperator {
```
Comment (acquamarin) [inline src/include/processor/operator/transaction.h:9]: Rename this class to TransactionOperation?

File: src/main/client_context.cpp [hunk 2]
```diff
@@ -17,10 +20,14 @@ void ActiveQuery::reset() {
...
-ClientContext::ClientContext()
+ClientContext::ClientContext(Database* database)
     : numThreadsForExecution{std::thread::hardware_concurrency()},
       timeoutInMS{ClientContextConstants::TIMEOUT_IN_MS}, varLengthExtendMaxDepth{
-                                                              VAR_LENGTH_EXTEND_MAX_DEPTH} {}
+                                                              VAR_LENGTH_EXTEND_MAX_DEPTH} {
+    transactionContext = std::make_unique<TransactionContext>(database);
+}
+
+ClientContext::~ClientContext() {}
```
Comment (acquamarin) [inline src/main/client_context.cpp:30]: If the deconstructor is empty, just let the compiler generates the default one.

File: src/main/connection.cpp [hunk 3]
```diff
@@ -27,36 +25,25 @@ namespace main {
...
 
-Connection::~Connection() {
-    if (activeTransaction) {
-        database->transactionManager->rollback(activeTransaction.get());
-    }
-}
+Connection::~Connection() {}
 
 void Connection::beginReadOnlyTransaction() {
-    std::unique_lock<std::mutex> lck{mtx};
```
Comment (acquamarin) [inline src/main/connection.cpp:41]: Can you add a comment that those apis are deprecated. Developers should avoid using these

File: src/main/prepared_statement.cpp [hunk 4]
```diff
@@ -13,6 +13,10 @@ bool PreparedStatement::allowActiveTransaction() const {
     return !StatementTypeUtils::allowActiveTransaction(preparedSummary.statementType);
 }
 
+bool PreparedStatement::isTransactionStatement() const {
```
Comment (acquamarin) [inline src/main/prepared_statement.cpp:16]: inline this

File: src/transaction/transaction_context.cpp [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    if (!allowActiveTransaction) {
+        throw ConnectionException(
+            "DDL, Copy, createMacro statements can only run in the AUTO_COMMIT mode. Please commit "
+            "or rollback your previous transaction if there is any and issue the query without "
+            "beginning a transaction");
+    }
+}
+
+void TransactionContext::commit() {
+    commitInternal(false);
```
Comment (acquamarin) [inline src/transaction/transaction_context.cpp:53]: Add comment for false, true

**Review by acquamarin — APPROVED** [review]

**Review by ray6080 — APPROVED** [review]

File: src/include/binder/visitor/statement_read_write_analyzer.h [hunk 6]
```diff
@@ -7,11 +7,19 @@ namespace binder {
 
 class StatementReadWriteAnalyzer : public BoundStatementVisitor {
 public:
-    StatementReadWriteAnalyzer() : BoundStatementVisitor(), readOnly{false} {}
+    StatementReadWriteAnalyzer() : BoundStatementVisitor(), readOnly{true} {}
 
     bool isReadOnly(const BoundStatement& statement);
 
 private:
+    void visitCreateTable(const BoundStatement& statement) override { readOnly = false; }
```
[the remark this answers is not in the document]
Comment (andyfengHKU) [reply src/include/binder/visitor/statement_read_write_analyzer.h:15]: will do this soon. Though this is not as easy as adding visitDDL here. It require us to unify the statementType to one.

File: src/main/client_context.cpp (hunk 2 above)
Comment (andyfengHKU) [reply src/main/client_context.cpp:30]: We are forward declaring. So cannot apply default constructor in header.

File: src/main/prepared_statement.cpp (hunk 4 above)
Comment (andyfengHKU) [reply src/main/prepared_statement.cpp:16]: ditto fwd declaring.

File: src/main/connection.cpp (hunk 3 above)
Comment (andyfengHKU) [reply src/main/connection.cpp:41]: Added an issue

**Pull request merged**